### ISSUE TITLE
Missing function for mask checking if mask is empty/mixed/full

### ISSUE THREAD

**Issue by noma** [edanor/umesimd#8]
I am missing an equivalent to Vc's `isEmpty()` function on the masks, or maybe a conversion operator to bool, that is true if at least one element of the mast is true.

With Intel you could use:

```
_mm512_mask2int
```

For other compilers, I use this replacement:

```
#if !defined(__INTEL_COMPILER)
// online defined by Intel Compiler
inline int _mm512_mask2int(__mmask16 k1) {
    return static_cast<int>(k1);
}
#endif
```

Rationale: You usually have three cases, that might be beneficial to check and handle differently:
- empty mask: nothing to do
- mixed mask: proceed with masked operation
- full mask: proceed with unmasked operations

**Comment by edanor** [comment]
You have to start thinking about mask as packed 'bool' values instead of some integer representation:

- empty mask: a mask is empty (all false) when 'logical or' of all elements is 'false'. Use Horizontal Logical OR operation (HLOR, called as .hlor()) and compare to 'false'.
- full mask: a mask is full (all true) when 'logical and' of all elements is 'true'. Use Horizontal Logical AND operation (HLAND, called as .hland()) and compare to 'true'.
- mixed mask:  a mask is mixed (some values 'true', some 'false') when it is neither 'empty' nor 'full'. So writing: !x.hlor() || x.hland())

In other words:
if(!mask.hlor()) // empty mask
if(mask.hland()) // full mask
if((!mask.hland()) && (mask.hlor())) // mixed mask

I can agree about the last one needing a dedicated instruction. The rest should already be there.

**Comment by noma** [comment]
Thanks, that works for now.

That's exactly how I think about masks, the integer representation is just very handy to efficiently implement the `isEmpty()` check, not sure how it compares to the `hlor()` implementation.

However, I missed the horizontal thing when searching through the code for an appropriate operations. Maybe that's just me, but I find the names in Vc a bit more intuitive as they reflect a higher level of abstraction.

**Comment by edanor** [comment]
> I missed the horizontal thing when searching through the code for an appropriate operations

All the operations are defined by design in UMESimdInterface.h. At least for now this is the only definitive and reliable reference.

The implementation for avx512 is just what you explained before. For other architectures it might be more complicated. Using horizontal reduction makes the concept closer to what we know from scalar programming, and there are already reductions for arithmetic types. So it should make thing a little bit easier for people who are just starting with SIMD programming.

**Comment by edanor** [comment]
Just checked out and I added an instruction that should behave the way you need (without further extensions to the interface). The mnemonic is LANDNOTV and it is a part of the masking interface. There is also a bitwise version of this available for integer types (BANDNOTV).
